Upgrade to lestrrat-go/jwx/v3

kw510 opened the pull request:
- Update dependencies to use v3 of lestrrat-go/jwx and related packages

- Refactor JWT handling in tests and main code for compatibility with new API.

kw510 committed:
Upgrade to lestrrat-go/jwx/v3 and apply migration changes

plutov commented:
Yes please, this brings issues!

plutov commented:
@kw510 did you manage to make tests work? I did similar PR, but `jwt.Sign` complains. https://github.com/go-chi/jwtauth/pull/102

kw510 commented:
@plutov  Heya! Just had a look and you had the same issue I had, but i did get the tests working in the end 🙈 

I fixed it by this replacement :)
```diff
- // NewSignatureAlgorithm creates a new SignatureAlgorithm object with the given name.
- jwa.NewSignatureAlgorithm
+ // LookupSignatureAlgorithm returns the SignatureAlgorithm object for the given name.
+ jwa.LookupSignatureAlgorithm
```

kw510 committed:
Update dependencies and refactor JWT claim extraction

- Upgrade Go version to 1.24.3
- Update lestrrat-go/jwx/v3 to version 3.0.2
- Upgrade golang.org/x/crypto to version 0.38.0 and golang.org/x/sys to version 0.33.0
- Refactor JWT claim extraction in jwtauth.go and jwtauth_test.go to use transform.AsMap for improved readability and error handling.

kw510 committed:
Update CI workflow to support Go versions 1.22.x, 1.23.x, and 1.24.x

plutov commented:
@kw510 please also change the version in the example file.

kw510 committed:
Update example file to use v3 lib

VojtechVitek commented on go.mod:
```diff
@@ -1,12 +1,10 @@
-go 1.23.0
-
-toolchain go1.24.2
+go 1.24.3
```
Can you please downgrade this back to Go 1.23.0?

This change would force all users of jwtauth to upgrade to 1.24.3. We want to keep some backward compatibility.

VojtechVitek commented on jwtauth_test.go:
```diff
@@ -1,3 +1,4 @@
+//go:debug rsa1024min=0
```
I'm just reading up on this -- https://go.dev/doc/godebug.

> Go 1.24 changed [crypto/rsa](https://go.dev/pkg/crypto/rsa) to require RSA keys to be at least 1024 bits. This behavior can be controlled with the rsa1024min setting. Using rsa1024min=0 restores the Go 1.23 behavior.

Sounds like we should fix the RSA keys instead?

VojtechVitek reviewed (commented):
Looks good! Minor feedback

Thank you 👍

kw510 committed:
Update Go version to 1.23.0 and modify RSA keys in tests

- Downgrade Go version in go.mod to 1.23.0
- Update RSA private and public keys in jwtauth_test.go for improved security and compatibility
- Adjust token string in TestSimpleRSAVerifyOnly for consistency

plutov commented:
@VojtechVitek @kw510 let's merge it ? ;)

VojtechVitek reviewed (approved):
LGTM. Thank you.

Let me test this some more before we merge 👍 

Can anyone else help test this PR?

plutov commented:
Hey @VojtechVitek I tested it on a project where it was originally failing, but now it works well. Looks good from my side!

plutov commented:
@VojtechVitek how is testing going? Why this merge takes so long?

hacdias commented:
Same question as @plutov: I've been holding off updating some libraries because of this PR. What's the status?

kosmoz commented:
Any progress on this? Can we help?

orchowski commented:
hey folks, @kw510 , @VojtechVitek , @plutov has this been tested already? It’s starting to become a bit of a pain. If you have any updates or a rough plan for when it might be released, I’d really appreciate it.

VojtechVitek commented:
@kosmoz @orchowski @hacdias can you help me test this change?

`go get github.com/go-chi/jwtauth/v5@c7560cd7521b4c40d025cb6254f327f4009cef17`

Once confirmed, I'll go ahead and merge.

kosmoz commented:
> @kosmoz @orchowski @hacdias can you help me test this change?
> 
> `go get github.com/go-chi/jwtauth/v5@c7560cd7521b4c40d025cb6254f327f4009cef17`
> 
> Once confirmed, I'll go ahead and merge.

I had to add this to my `go.mod` file:

```
replace github.com/go-chi/jwtauth/v5 => github.com/kw510/jwtauth/v5 v5.0.0-20250603171652-c7560cd7521b
```

I did a quick test on https://github.com/distr-sh/distr and it worked (after updating jwx to v3, obviously)

hacdias commented:
Seems to work!

orchowski commented:
Also works

orchowski commented:
@VojtechVitek any more blockers?

VojtechVitek reviewed (approved):
LGTM!

Great, thanks everyone for your help with reviews & testing this PR.

@kw510 Thank you for your contribution and for your patience.
